Task2: Construction of a deterministic finite automaton by a regular expression and a non-deterministic finite automaton by a graph

SurfaceYellowDuck opened the pull request:
Add regex_to_dfa and graph_to_nfa function, add tests for synthetic graph and graph from file, enabled autotests for second task.

SurfaceYellowDuck committed:
Add regex_to_dfa and graph_to_nfa functions
add tests for synthetic graph and graph from file
enabled autotests

KubEF commented on project/hw2/graph_to_nfa_tool.py:
```diff
@@ -0,0 +1,46 @@
+import networkx
+
+
+@dataclass
+class Graph:
+    edges_cnt: int
+    nodes_cnt: int
+    labels: Set
+
+
+def load_graph(graph_name: str):
+    path = cfpq_data.download(graph_name)
+    graph = cfpq_data.graph_from_csv(path)
+    edges_cnt = graph.number_of_edges()
+    nodes_cnt = graph.number_of_nodes()
+    labels = cfpq_data.get_sorted_labels(graph)
+    return Graph(edges_cnt, nodes_cnt, set(labels))
+
+
+def build_two_cycle_graph(n: int, m: int, labels: Tuple[str, str], path="./test2.dot"):
+    graph = cfpq_data.labeled_two_cycles_graph(n, m, labels=labels)
+    pydot_graph = networkx.drawing.nx_pydot.to_pydot(graph)
+    pydot_graph.write_raw(path)
```
Это зачем? У вас это есть в первой домашке, зачем переписывать?

SurfaceYellowDuck replied:
Я там dataclass поменял, labels теперь set, а не список. Ну и для того, чтобы из первой домашки методы не тащить при тестировании

SurfaceYellowDuck replied:
Надо ли это убирать, и переделывать код из первой домашки?

KubEF replied:
Ну, оставить одну реализацию: ту, которая была в первой домашке. Я не очень понимаю, что там протаскивать. Но да, если что-то надо импортировать -- импортируйте. На каждый чих всё переписывать очень-очень странно

SurfaceYellowDuck replied:
У вас по условиям код из прошлых домашек не может оказаться в коммитах с кодом к текущей домашке, при этом для того чтобы тут тесты проходили мне надо поменять тип labels в датаклассе с list на set. Будет ли какое-то наказание, если этот фикс всё-таки окажется здесь в коммите?

KubEF replied:
Это дискуссионный вопрос. Сейчас у вас половина баллов и так слетело, так что не супер важно, но в следующий раз не стесняйтесь написать в чатик, он для этого нужен как раз. Но вот переписывать предыдущую домашку в этой полнстью -- совсем странное решение. Поправьте код первой домашки как вам будет удобно.

KubEF commented on project/hw2/graph_to_nfa_tool.py:
```diff
@@ -0,0 +1,46 @@
+) -> NondeterministicFiniteAutomaton:
+    nfa = NondeterministicFiniteAutomaton.from_networkx(graph)
+    all_states = nfa.states
+    if not start_states or not final_states:
+        for st in all_states:
+            nfa.add_start_state(st)
+            nfa.add_final_state(st)
```
Вы уверены, что это должно работать так? Условие про отсутствте финальных и стартовых нужно просто чтобы можно было не передавать `set(graph.nodes)` всякий раз, когда хочется сделать финальными или стартовыми все. А это уже какое-то дополнительное непредсказуемое поведение.

Но спасибо за нахождение дыры в наших автотестах.

SurfaceYellowDuck replied:
Вы не передавать не сможете, вам сигнатура функции не позволит. Хотя бы пустой set передать придется

SurfaceYellowDuck replied:
Ситуация, когда передали не пустым только final_states или только start_states является корректной?

SurfaceYellowDuck committed:
Fix: changed the behavior of the graph_to_nfa function
Removed the redundant code, replaced it with imports
Add tests in which only initial or only final states are passed to the function

KubEF commented on project/hw2/graph_to_nfa_tool.py:
```diff
@@ -0,0 +1,34 @@
+) -> NondeterministicFiniteAutomaton:
+    nfa = NondeterministicFiniteAutomaton.from_networkx(graph)
+    all_states = nfa.states
+    if not start_states and not final_states:
+        for st in all_states:
+            nfa.add_start_state(st)
+            nfa.add_final_state(st)
+
+    elif not start_states and final_states:
+        for st in final_states:
+            nfa.add_final_state(State(st))
+        for st in all_states:
+            nfa.add_start_state(st)
+
+    elif not final_states and start_states:
+        for st in start_states:
+            nfa.add_start_state(State(st))
+        for st in all_states:
+            nfa.add_final_state(st)
+
+    else:
+        for st in start_states:
+            nfa.add_start_state(State(st))
+        for st in final_states:
+            nfa.add_final_state(State(st))
+    return nfa
```
Уф. Это формально правильно, так что я зачту, но советую посмотреть, как это нормально можно сделать.
